src/realm/parser/query_builder.cpp

```diff
@@ -36,131 +36,129 @@ using namespace query_builder;
+template <>
+Query do_add_null_comparison_to_query<Link>(Predicate::Operator op, const PropertyExpression& expr)
 {
     REALM_ASSERT_DEBUG(!expr.dest_type_is_backlink()); // this case should be handled at a higher layer
     switch (op) {
-        case Predicate::Operator::NotEqual:
-            query.Not();
-            REALM_FALLTHROUGH;
+        case Predicate::Operator::NotEqual: {
+            Columns<Link> column = expr.value_of_type_for_query<Link>();
+            return column != realm::null();
+        }
         case Predicate::Operator::In:
-            REALM_FALLTHROUGH;
         case Predicate::Operator::Equal: {
             Columns<Link> column = expr.value_of_type_for_query<Link>();
-            query.and_query(column == realm::null());
-            break;
+            return column == realm::null();
         }
         default:
-            throw std::logic_error("Only 'equal' and 'not equal' operators supported for object comparison.");
+            throw_logic_error("Only 'equal' and 'not equal' operators supported for object comparison.");
     }
 }
 
 // add a clause for numeric constraints based on operator type
 template <typename A, typename B>
-void add_numeric_constraint_to_query(Query& query,
-                                     Predicate::Operator operatorType,
-                                     A lhs,
-                                     B rhs)
+Query add_numeric_constraint_to_query(Predicate::Operator operatorType, A lhs, B rhs)
 {
     switch (operatorType) {
         case Predicate::Operator::LessThan:
-            query.and_query(lhs < rhs);
-            break;
+            return lhs < rhs;
         case Predicate::Operator::LessThanOrEqual:
-            query.and_query(lhs <= rhs);
-            break;
+            return lhs <= rhs;
         case Predicate::Operator::GreaterThan:
-            query.and_query(lhs > rhs);
-            break;
+            return lhs > rhs;
         case Predicate::Operator::GreaterThanOrEqual:
-            query.and_query(lhs >= rhs);
-            break;
+            return lhs >= rhs;
         case Predicate::Operator::In:
-            REALM_FALLTHROUGH;
         case Predicate::Operator::Equal:
-            query.and_query(lhs == rhs);
-            break;
+            return lhs == rhs;
         case Predicate::Operator::NotEqual:
-            query.and_query(lhs != rhs);
-            break;
+            return lhs != rhs;
         default:
-            throw std::logic_error("Unsupported operator for numeric queries.");
+            throw_logic_error("Unsupported operator for numeric queries.");
     }
 }
 
 template <typename A, typename B>
-void add_bool_constraint_to_query(Query &query, Predicate::Operator operatorType, A lhs, B rhs) {
+Query add_bool_constraint_to_query(Predicate::Operator operatorType, A lhs, B rhs)
+{
     switch (operatorType) {
         case Predicate::Operator::In:
-            REALM_FALLTHROUGH;
```

Comment:
We may not be tracking these types of warnings right now, but I at least found this macro to be a useful indication that the the fallthrough is not a programmatic error. Do you have a reason for removing it?

Reply:
I think the warning is on by default with -Wall. However you only need to annotate when there is code between the cases, not when cases are directly stacked like this. I kinda like having the fall through annotation be a signal that there is "interesting" control flow going on, not just multiple cases handled by identical code. 

That said, if you found them helpful in these cases, I can put them back.

Reply:
> However you only need to annotate when there is code between the cases, not when cases are directly stacked like this.

I must have forgotten this. There is no style convention for this that I know of and your reasoning makes sense to me, so I'm fine with the removal in the trivial cases. Thanks for explaining!

Reply:
BTW, while it isn't obvious due to the trivial cases like this, this patch *does* remove some cases of "interesting" fallthrough (not just removing the macro 😉), such as in `do_make_null_comparison_query` and ` make_link_constraint_query`. It is probably worth reviewing those a bit more closely.